src/mediapipe_internal/mediapipegraphexecutor.cpp

```diff
@@ -952,23 +952,31 @@ Status MediapipeGraphExecutor::validateSubsequentRequest(const ::inference::Mode
     return StatusCode::OK;
 }
 
+static bool streamSynchronizedWrite(::grpc::ServerReaderWriterInterface<::inference::ModelStreamInferResponse, KFSRequest>& stream,
+    std::mutex& mtx, ::inference::ModelStreamInferResponse& resp) {
+    const std::lock_guard<std::mutex> lock(mtx);
+    return stream.Write(resp);
+}
+
 Status MediapipeGraphExecutor::inferStream(const KFSRequest& firstRequest, ::grpc::ServerReaderWriterInterface<::inference::ModelStreamInferResponse, KFSRequest>& stream) {
     SPDLOG_DEBUG("Start streaming KServe request mediapipe graph: {} execution", this->name);
     try {
         // Init
         ::mediapipe::CalculatorGraph graph;
         MP_RETURN_ON_FAIL(graph.Initialize(this->config), "graph initialization", StatusCode::MEDIAPIPE_GRAPH_INITIALIZATION_ERROR);
 
+        std::mutex streamWriterMutex;
```

Comment:
Is it guaranteed that the live scope of this mutex will outlive the scope of the graph? I guess this is ok if we would reach the WaitUntilDone(), but clearly we could have some kind of exception and in that case if  we won't wait until graph is done, one of the graph threads could use callback while other thread worker thrown exception?

Especially given that mutex is destroyed earlier on than graph.

Reply:
I'll move mutex creation to the beginning of that function, so graph gets destroyed first.